### DISCUSSION TITLE
Stream Playlists on Twitch, Youtube or any RTMP Endpoint Smoothly

### DISCUSSION THREAD

**Discussion by mekya** [ant-media/Ant-Media-Server#6158] [Show and tell]
Hi Guys,
As shared previously, there are instructions how to have a smooth playlist playback experience (without freezes) on Ant Media Server. To learn these instructions, check this post: [Seamless Playlist Experience in Ant Media Server](https://medium.com/@ahmetmermerkaya/seamless-playlist-experience-in-ant-media-server-bbd73f7a70e6). You can also check this discussion: [Basic PlayList Guide From Zero to Hero](https://github.com/orgs/ant-media/discussions/6015)

Today, we're going to provide a workaround solution to Stream PlayLists on Twitch Smoothly because when we directly add RTMP endpoint to the playlist, streaming has restarted on Twitch whenever playlist switches items.  I think this is an important topic because there are multiple discussions([Playist always start a new stream...](https://github.com/orgs/ant-media/discussions/5364), [Stream Playlist VOD on Twitch](https://github.com/orgs/ant-media/discussions/4977)) and issues([Streaming a playlist to twitch does not work correctly](https://github.com/ant-media/Ant-Media-Server/issues/5005),  [The playlist freezes and .ts files are missing for uploaded VODs.](https://github.com/ant-media/Ant-Media-Server/issues/4468)) around this problem. 

Here is the step by step instructions.

> Pay attention that this solution works for versions 2.8.3 and above. Currently, we've not released 2.8.3 so you can use the SNAPSHOT versions.  Community Edition SNAPSHOT is available on[ sonatype](https://oss.sonatype.org/#nexus-search;gav~io.antmedia~ant-media-server~~~~kw,versionexpand) and Enterprise Edition SNAPSHOT is available on your account on [antmedia.io](https://antmedia.io)    

# How to Stream Playlist on Twitch Smoothly

> If you already have experience about playlist on Ant Media Server, you can skip to Step 3 directly.

Tune the settings for having a smooth playlist experience
## Step 1: Configure Ant Media Server for Better Playlist Experience

If you know how to configure Ant Media Server for smooth playlist experience, you can skip to next section.

1. Login to your Ant Media Server Web Panel in. The url is like this `https://YOUR_ANT_MEDIA_SERVER:5443/`
<br/><br/><img width="700" alt="Screenshot 2024-01-05 at 10 34 35" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/1b82fc48-25b9-4815-a8d5-845704770113"><br/><br/>

2. Click the LiveApp on the left 
<br/><br/><img width="700" alt="Screenshot 2024-01-05 at 10 36 20" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/e68a14e3-e1e7-4332-aa21-a7409cda3011"><br/><br/>

3. Click the`Settings` tab and scroll bottom 
<br/><br/><img width="700" alt="Screenshot 2024-01-05 at 10 40 01" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/04a712ad-2cd9-4e01-8963-0e793d3209ac">
<br/><br/>

4. Uncheck the `Delete HLS files after the stream is finished`
<br/><br/><img width="700" alt="Screenshot 2024-01-08 at 12 52 05" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/aecbec31-8350-4015-9edc-e07958f120ee"><br/><br/>

5. Scroll bottom and Click `Save` button
<br/><br/><img width="700" alt="Screenshot 2024-01-08 at 12 56 25" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/c89361a8-e710-4ead-bebc-eed3e4d977d6"><br/><br/>

6. Scroll top and Choose the  mode from `Basic` to `Advanced`
<br/><br/><img width="700" alt="Screenshot 2024-01-08 at 12 58 14" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/e8a74022-2fa7-4cec-bc8e-4c70623085bf"><br/><br/>

7. Find `hlsflags` and change its value to  `delete_segments+append_list+omit_endlist`
<img width="700" alt="Screenshot 2024-01-08 at 13 00 57" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/f064e15b-8f51-4ce1-83d6-afa65b8e1f34">

## Step 2: Upload Some VoD Files and Create Your Playlist

If you know how to upload VoD files  and create playlist, you can skip to next section.

### Upload VoD Files
1. Click the `VoD` Tab for LiveApp and then click the `Upload VoD` button
 <br/><br/><img width="700" alt="Screenshot 2024-01-05 at 10 58 52" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/b98b52a5-0e94-4807-841b-cf72d78b4b8e"><br/><br/>

2. Upload some MP4 VoD Files from your computer
<br/><br/><img width="700" alt="Screenshot 2024-01-05 at 10 58 32" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/781fdb3b-4cdd-4115-81ca-8e7cab002778"><br/><br/>

3. We've added 3 VoD files
<br/><br/><img width="700" alt="Screenshot 2024-01-05 at 11 09 25" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/8f6ad953-1b8c-4159-82d7-b8518561ccd5"><br/><br/>

4. Copy the VoD Urls through context menu. We'll use them below. 
<img width="700" alt="Screenshot 2024-01-08 at 13 08 47" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/7a5615cd-da72-4912-be9c-dfcb3269bc4c">

So my VoDs are accessible through this links and I'll use them below. 
`http://localhost:5080/LiveApp/streams/243170266880429550645296.mp4`
`http://localhost:5080/LiveApp/streams/794333965265876491918328.mp4`
`http://localhost:5080/LiveApp/streams/823913410113173360449884.mp4`

### Create Playlist
1.  Click `Live Streams` Tab and then Click `New Live Stream > Playlist` as shown below
<br/><br/>
<img width="700" alt="Screenshot 2024-03-05 at 09 46 50" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/97eb4e2b-58d3-4261-b278-0c08978a8289">
<br/><br/>
2. Add the VoD files to your PlayList and Click the `Create` button. We've entered `myplaylist` as `streamId`
<br/><br/>
<img width="700" alt="Screenshot 2024-03-05 at 09 46 50" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/17592913-f730-4866-bffe-e2a1df20b873">
<br/><br/>
3. Click the "Start Broadcast" button to let the playlist start
<br/><br/>
<img width="700" alt="Screenshot 2024-03-05 at 09 48 57" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/3ee5a327-71e9-49b9-aa5e-8504b042ae38">
<br/><br/>

## Step 3: Stream Playlist on Twitch Smoothly
We've created our playlist with `myplaylist` id and it means that the stream is playable directly with the following url: `http://{YOUR_ANT_MEDIA_SERVER}:5080/LiveApp/streams/myplaylist.m3u8`

1. Create `Stream Source` 
<br/><br/>
<img width="700" alt="Screenshot 2024-03-05 at 09 59 07" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/877b00f3-f6c7-4e8e-934c-c398f2a5094f">
<br/><br/>
2. Use the playlist as Stream Source URL. We use `http://localhost:5080/LiveApp/streams/myplaylist.m3u8` as Stream Source URL because server can access it through local address.  
<br/><br/>
<img width="700" alt="Screenshot 2024-03-05 at 10 00 09" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/a949f3d5-89cd-427f-9ea5-197df9d5ff7a">
<br/><br/>
3.  Click `Edit RTMP Endpoints` 
<br/><br/>
<img width="700" alt="Screenshot 2024-03-05 at 10 00 32" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/4b60640f-cb1b-4e02-a51b-79402bc9f10a">
<br/><br/>
4. Add your Twitch Channel RTMP Endpoint. Use the RTMP server that is close to you from [ingest.twitch.tv/ingests](https://ingest.twitch.tv/ingests) and use your stream key from your account on twitch.
<br/><br/>
<img width="700" alt="Screenshot 2024-03-05 at 10 10 28" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/92a836ce-bdfa-4f8a-8804-919720077dc6">
<br/><br/>
5. Last step "Click Start Broadcasting" and check that Stream is playing without any interruption on Twitch. I've used one of my colleagues 7 seconds video(@burak-58 ) to check if any interruptions occurs between switches and the synchronization
<br/><br/>
<img width="700" alt="Screenshot 2024-03-05 at 10 36 52" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/367f774b-4447-46be-8f4a-3dbac21b38ea">

Please try and let me know if it works for you. 

Bonus: Twitch reports that Live stream is unstable. It's expected because of nature of HLS streaming.  I think it will be resolved if we  create the Stream Source as VoD for HLS stream programmatically. I've not tried yet. If anyone try and let me know it will perfect or I can try in one of my convenient time

**Comment by JustinoOn60hz** [comment]
I tried this and for some reason when I look at twitch it just shows a black screen instead of the actual video is there any fix to this?

**Reply by mekya** [reply]
Hi @JustinoOn60hz,

Thank you for your question. 

Is there a chance that you let us try some videos and see the problem the same way you see? Because I've not encounter this problem before.

Regards
Oguz

**Reply by guiiicd** [reply]
I had the same problem

**Reply by Mohit-3196** [reply]
Hi @guiiicd,
Thank you for reporting.

As @mekya mentioned above, ss there a chance that you let us try some videos and see the problem the same way you see it?
Also, let me know the AMS version that you are running.

Thank you,
Mohit

**Reply by guiiicd** [reply]
I updated to version 2.9.0 and it stopped giving me this error

**Reply by Mohit-3196** [reply]
Great. Thank you for letting us know @guiiicd

**Reply by spjoes** [reply]
Im having the same black screen issue on v2.10.0. Black screen no video but there is audio. Other times, it just shows as if the stream is offline but twitch has the "LIVE" symbol

**Reply by verasativa** [reply]
> Im having the same black screen issue on v2.10.0. Black screen no video but there is audio. Other times, it just shows as if the stream is offline but twitch has the "LIVE" symbol

Exactly the same here, black screen, "LIVE" icon, and Community Edition 2.10.0 20240705_1023.

Any ideas?

**Comment by tmbailey0304** [comment]
Hey there, the stream runs smoothly but after a couple of hours the stream becomes jittery and shakey. I am streaming to Twitch. Here is a little preview of how it looks. I assume this is because it is broadcasting at 0.69x but I am not sure why this would be happening. Any help on this matter would be greatly appreciated.

![image](https://github.com/ant-media/Ant-Media-Server/assets/103694864/3e71566c-820e-4599-8fbd-80a744b9543b)
https://github.com/ant-media/Ant-Media-Server/assets/103694864/97d16401-9477-4df6-a309-830a252febf1

**Reply by mekya** [reply]
Hi @tmbailey0304 ,

Sorry for missing this question, could you provide the videos to us? We may try and see the same problem again. It will help us a lot understand and fix the problem

Cheers
Oguz
